File: 445. Add Two Numbers II/Java/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        class Utils {
            int twoNumbers(ListNode l1, ListNode l2, int length1, int length2, ListNode result) {
                ListNode node = new ListNode(0);
                result.next = node;
                int sumTwo = 0;
                if (l1.next == null) {
                    sumTwo = l1.val + l2.val;
                } else if (length1 > length2) {
                    int carry = twoNumbers(l1.next, l2, length1 - 1, length2, result.next);
                    sumTwo = l1.val + carry;
                } else {
                    int carry = twoNumbers(l1.next, l2.next, length1 - 1, length2 - 1, result.next);
                    sumTwo = l1.val + l2.val + carry;
                }
                node.val = sumTwo % 10;
                return sumTwo / 10;
            }
        }
        int length1 = 0;
        ListNode temp = l1;
        while (temp != null) {
            length1++;
            temp = temp.next;
        }
        int length2 = 0;
        temp = l2;
        while (temp != null) {
            length2++;
            temp = temp.next;
        }
        ListNode result = new ListNode(1);
        Utils utils = new Utils();
        int carry = length1 > length2 ? utils.twoNumbers(l1, l2, length1, length2, result) : utils.twoNumbers(l2, l1, length2, length1, result);
        if (carry == 0) {
            result = result.next;
        }
        return result;
    }
}

File: 445. Add Two Numbers II/Python/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        length_1 = 0
        temp = l1
        while temp:
            length_1 += 1
            temp = temp.next
        length_2 = 0
        temp = l2
        while temp:
            length_2 += 1
            temp = temp.next

        def two_numbers(l1, l2, length_1, length_2, result):
            node = ListNode(0)
            result.next = node
            if not l1.next:
                sum_two = l1.val + l2.val
            elif length_1 > length_2:
                carry = two_numbers(l1.next, l2, length_1 - 1, length_2, result.next)
                sum_two = l1.val + carry
            else:
                carry = two_numbers(l1.next, l2.next, length_1 - 1, length_2 - 1, result.next)
                sum_two = l1.val + l2.val + carry
            node.val = sum_two % 10
            return sum_two // 10

        result = ListNode(0)
        carry = two_numbers(l1, l2, length_1, length_2, result) if length_1 > length_2 else two_numbers(l2, l1,
                                                                                                        length_2,
                                                                                                        length_1,
                                                                                                        result)
        if carry:
            result.val = 1
        else:
            result = result.next
        return result
